### PR TITLE
Extend product with Docker, Puppet, OSTree repos

### PR DESCRIPTION
**Pull request by ldjebran**
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v tests/foreman/cli/test_vm_install_products_package.py 
============================================ test session starts =============================================
platform linux -- Python 3.6.6, pytest-4.0.2, py-1.5.4, pluggy-0.8.0 -- /home/dlezz/.pyenv/versions/robottelo/bin/python3.6
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.26.0, timeout-1.3.3, services-1.3.1, mock-1.10.0, forked-0.2, cov-2.6.1
collecting ... 2019-03-05 15:41:11 - conftest - DEBUG - BZ deselect is disabled in settings

collected 4 items                                                                                            

tests/foreman/cli/test_vm_install_products_package.py::test_vm_install_package[0] PASSED               [ 25%]
tests/foreman/cli/test_vm_install_products_package.py::test_vm_install_package[1] PASSED               [ 50%]
tests/foreman/cli/test_vm_install_products_package.py::test_vm_install_package[2] PASSED               [ 75%]
tests/foreman/cli/test_vm_install_products_package.py::test_vm_install_package[3] PASSED               [100%]

================================== 4 passed, 2 warnings in 2094.25 seconds ===================================
```

### CODE REVIEW

File: robottelo/products.py [hunk 1]
```diff
@@ -316,14 +363,59 @@ class DockerRepository(BaseRepository):
...
 
+    def __init__(self, url=None, distro=None, upstream_name=None):
+        self._upstream_name = upstream_name
+        super(DockerRepository, self).__init__(url=url, distro=distro)
+
+    @property
+    def upstream_name(self):
+        return self._upstream_name
+
+    def create(self, organization_id, product_id, download_policy=None, synchronize=True):
```
Comment (oshtaier) [inline robottelo/products.py]: any reason to override that method? Can't we implement something like:
```
if self.content_type == REPO_TYPE_DOCKER:
    create_options['download-policy'] = self.upstream_name
```

File: robottelo/products.py [hunk 2]
```diff
@@ -654,14 +779,97 @@ def setup(self, org_id, download_policy=DOWNLOAD_POLICY_ON_DEMAND,
...
+                'organization-id': org_id,
+                'to-lifecycle-environment-id': lce_id,
+            })
+        content_view = ContentView.info({u'id': content_view['id']})
+        return content_view, lce
+
+    @staticmethod
+    def setup_activation_key(org_id, content_view_id, lce_id, subscription_names=None):
+        # type: (int, int, int, Optional[List[str]], Optional[str]) -> Dict
+        """"""
```
Comment (abalakh) [inline robottelo/products.py]: missing docstring

File: robottelo/products.py [hunk 3]
```diff
@@ -654,14 +779,97 @@ def setup(self, org_id, download_policy=DOWNLOAD_POLICY_ON_DEMAND,
...
+        if any(not repo.cdn for repo in self):
+            custom_product = make_product_wait({
+                'organization-id': org_id,
+            })
+        custom_product_id = custom_product['id'] if custom_product else None
+        for repo in self:
+            repo_info = repo.create(org_id, custom_product_id,
+                                    download_policy=download_policy, synchronize=synchronize)
+            repos_info.append(repo_info)
+
```
Comment (abalakh) [inline robottelo/products.py]: nitpick: redundant empty line

File: robottelo/products.py [hunk 4]
```diff
@@ -654,14 +779,97 @@ def setup(self, org_id, download_policy=DOWNLOAD_POLICY_ON_DEMAND,
...
-        setup_data = setup_cdn_and_custom_repositories(
-            org_id,
-            self.repos_data,
-            download_policy=download_policy,
-            synchronize=synchronize,
+        custom_product = None
+        repos_info = []
+        if any(not repo.cdn for repo in self):
+            custom_product = make_product_wait({
+                'organization-id': org_id,
```
Comment (abalakh) [inline robottelo/products.py]: nitpick: fits in one line

File: robottelo/products.py [hunk 5]
```diff
@@ -478,6 +570,40 @@ def __repr__(self):
...
+        # type: (int, Optional[int], Optional[str], Optional[bool]) -> Dict
+        """Create an RH repository"""
+        if not self.cdn and not self.url:
+            raise ValueError(u'Can not handle Custom repository with url not supplied')
+        if self.cdn:
+            data = self.data
+            RepositorySet.enable({
+                'organization-id': organization_id,
+                'product': data['product'],
+                u'name': data['repository-set'],
```
Comment (abalakh) [inline robottelo/products.py]: nitpick: pls remove that `u` part from all strings.

File: robottelo/products.py [hunk 6]
```diff
@@ -654,14 +779,97 @@ def setup(self, org_id, download_policy=DOWNLOAD_POLICY_ON_DEMAND,
...
+            if (subscription['name'] in subscription_names
+                    and subscription['name'] not in added_subscription_names):
+                ActivationKey.add_subscription({
+                    'id': activation_key['id'],
+                    'subscription-id': subscription['id'],
+                    'quantity': 1,
+                })
+                added_subscription_names.append(subscription['name'])
+                if (len(added_subscription_names)
+                        == len(subscription_names)):
```
Comment (abalakh) [inline robottelo/products.py]: nitpick: fits in one line

File: robottelo/products.py [hunk 7]
```diff
@@ -717,14 +933,17 @@ def setup_virtual_machine(
         :param install_katello_agent: whether to install katello-agent
         :param enable_rh_repos: whether to enable RH repositories
         :param enable_custom_repos: whether to enable custom repositories
+        :param configure_rhel_repo: Whether to configures the distro Red Hat repository, this is
```
Comment (abalakh) [inline robottelo/products.py]:
```suggestion
        :param configure_rhel_repo: Whether to configure the distro Red Hat repository, this is
```

File: robottelo/products.py [hunk 8]
```diff
@@ -316,14 +363,59 @@ class DockerRepository(BaseRepository):
...
+
+    @property
+    def puppet_modules(self):
+        return self._puppet_modules
+
+    def add_to_content_view(self, organization_id, content_view_id):
+        options = {
+            'content-view-id': content_view_id,
+            'organization-id': organization_id,
+        }   # type: Dict[str, Any]
```
Comment (abalakh) [inline robottelo/products.py]: What is this type hint for?

File: robottelo/products.py [hunk 9]
```diff
@@ -316,14 +363,59 @@ class DockerRepository(BaseRepository):
...
+        # type: (str, Optional[str], List[Dict[str, str]]) -> None
+        self._puppet_modules = modules
+        super(PuppetRepository, self).__init__(url=url, distro=distro)
+
+    @property
+    def puppet_modules(self):
+        return self._puppet_modules
+
+    def add_to_content_view(self, organization_id, content_view_id):
+        options = {
```
Comment (abalakh) [inline robottelo/products.py]: missing docstring

File: robottelo/products.py (hunk 8 above)
Comment (ldjebran) [reply robottelo/products.py]: no just an old implementation and forget to remove :)

**Comment by ldjebran** [comment]
@abalakh changes done

**Pull request merged**